File: java/139-word-break.java
class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        if (s == null || s.length() == 0 || wordDict == null || wordDict.size() == 0) {
            return false;
        }
        int maxLength = getMaxLength(wordDict);
        Set<String> dict = new HashSet<>(wordDict);
        boolean[] dp = new boolean[s.length() + 1];
        dp[0] = true;
        for (int i = 1; i < dp.length; i++) {
            boolean isMatch = false;
            for (int j = Math.max(0, i - maxLength); j < i; j++) {
                if (dp[j] && dict.contains(s.substring(j, i))) {
                    isMatch = true;
                }
            }
            dp[i] = isMatch;
        }
        return dp[dp.length - 1];
    }
    private int getMaxLength(List<String> words) {
        int max = Integer.MIN_VALUE;
        for (String word : words) {
            max = Math.max(word.length(), max);
        }
        return max;
    }
}

File: python/139-word-break.py
class Solution:
    def wordBreak(self, s, wordDict):
        """
        :type s: str
        :type wordDict: List[str]
        :rtype: bool
        """
        if not wordDict or len(wordDict) == 0:
            return False
        if not s or len(s) == 0:
            return True
        dict = set(wordDict)
        # - get max length
        maxLength = 0
        for word in wordDict:
            maxLength = max(maxLength, len(word))
        dp = [False] * (len(s) + 1)
        dp[0] = True
        for i in range(1, len(dp)):
            start = max(0, i - maxLength)
            for j in range(start, i):
                if s[j:i] in dict and dp[j]:
                    dp[i] = True
        return dp[-1]
        